lib/route53.js

```diff
@@ -28,99 +69,46 @@ class AWSHandler extends DNSHandler {
 		return this._awsSdk;
 	}
 
-	_getHostedZoneId() {
+	getHostedZoneId() {
 		return this.Route53.listHostedZonesByName({
 			DNSName: this.domain,
 		}).promise()
-			.then(response => _.filter(_.flatMap(response), [
-				"Config.PrivateZone",
-				false,
-			])[0].Id)
-			.catch(error => {
-				console.log("[ERROR]", error);
-			});
-	}
-
-	_generateParams(hostedZoneId, opts) {
-		let params = {
-			HostedZoneId: hostedZoneId,
-			ChangeBatch: {
-				Changes: [
-					{
-						Action: "",
-						ResourceRecordSet: {
-							Name: opts.name.concat(".", this.domain, "."),
-							Type: opts.type || "CNAME",
-							TTL: opts.ttl || 3600,
-						},
-					},
-				],
-			},
-		};
-
-		if (opts.values) {
-			params.ChangeBatch.Changes[0].ResourceRecordSet.ResourceRecords = opts.values.map(function(r){
-				return {
-					Value: r
-				}
-			});
-		}
-		return Promise.resolve(params);
+		.then(extractZoneIdFromResponse)
+		.catch(console.error);
 	}
 
 	addRecord(opts) {
-		return this._getHostedZoneId()
-			.then(hostedZoneId => this._generateParams(hostedZoneId, opts))
-			.then((params) => {
-				params.ChangeBatch.Changes[0].Action = "UPSERT";
-				return Promise.resolve(params);
-			})
+		return this.getHostedZoneId()
+			.then(hostedZoneId => generateParams(hostedZoneId, this.domain, opts))
+			.then(params => changeRequestMethod(params, "UPSERT"))
 			.then(params => this.Route53.changeResourceRecordSets(params).promise())
```

Comment:
I would consider collapsing this one with the one above, since the `changeRequestMethod` is not really async. You're making it wait a tick (1ms) unnecessarily.

Reply:
e.g. the long "functional" way:

```javascript
.then(params => this.Route53.changeResourceRecordSets(changeRequestMethod(params, "UPSERT")).promise())
```

Reply:
You will need to pull the generateParams down as well since it is not async either.